### ISSUE TITLE
Research Write-Path Compression

### ISSUE THREAD

**Issue by andreipaduroiu** [pravega/pravega#4298]
### Problem description
Investigate the possibility of compressing the data in-flight or in Tier 1.

Goals:
- Tier 2 performance, Read performance and Recovery times should not be affected.
- Non-compressible data should have the same ingestion performance as without compression.
- Compressible data should show meaningful improvement in throughput and latencies.
- Completely transparent to the user.

Non-goals:
- Compressing data in Tier 2 or the cache.

### Problem location
Segment Store/Tier 1/Ingestion Pipeline.

### Suggestions for an improvement
**Where to compress?**
- In the Client (EventStreamWriter). This will move CPU burden outside of Pravega, however benefits would be small since we'd have to compress each event at a time. On the plus side, for those big events that really are compressible, the total amount of data sent over the wire is reduced significantly.
- In `BookKeeperLog`. Compress the buffers before sending them to BookKeeper. We have the entire buffer to analyze and compress so this will be the best compression we can ever get. The downside is that we may end up having to create a new buffer to store the compressed data, but it can be alleviated.

**When to compress?**
- Compressing works best when the following two conditions are met:
    - The data is compressible (Compressible means the compression ratio is < 80%).
    - It is faster to compress the data and write the compressed data than it is to write the original data (compression takes time, which is a factor of the size of the data and its complexity).
- We need a heuristic that can quickly determine if a specific data buffer is compressible in a small amount of time:
    - Ideally this would be near-instantaneous for non-compressible data, but that may not always be the case. It is ok for a solution that will immediately reject the vast majority of non-compressible data and take up to 1ms for the rest (which also includes compressible data).
    - The heuristic should have a high degree of accuracy (90%+ for both compressible and non-compressible data).

**Possible Implementation in BookKeeperLog**
This is based off a functioning prototype that has been measured using the SelfTester tool. See below for results.
- Compression is done in `BookKeeperLog`, immediately before sending the write to BK. This must happen asynchronously with respect to the `OperationProcessor`'s Ingestion Pipeline (i.e., in a separate thread).
- Compression is done only subject to the following heuristic:
    - **Buffer size** is greater than 512KB. 
        - The smaller the buffer, the least likely it will be that it's compressible. 
        - Smaller buffers are also an indication that BK is lightly loaded, so it can process a new write very quickly; the compression latency would actually make performance worse.
    - Recent BK **Write Latencies** are greater than double the recent **Compression Latencies**.
        - Size is not always a good indicator of compression latency. As such, we want to "guess" whether compressing the data and writing the compressed data is faster or slower than just writing the original data to BK.
        - We calculate a Simple Moving Average of the last few BK Writes (`B`) and a Simple Moving Average of the last few Compression calls (`C`). If `B < 2 * C`, then we should not compress (we multiply by `2` because `B` already includes previous values of `C`).
    - Sampling the buffer yields a **bi-gram entropy** less than 10.
        - Entropy is a good measure of determining the randomness of data. 
        - However single-byte entropy cannot differentiate between data that uses all possible byte values (255) randomly or in a pattern.
        - Considering bi-grams (adjacent byte pairs), we can calculate their entropy and get a better indication of whether the data buffer is likely to contain repeated patterns (and thus be compressible).
        - We need not check the whole buffer (1MB). Sampling random slices of it would do just fine, as long as we preserve statistical significance. For example, 8 samples of 16KB is sufficient for a 1MB buffer.

**Possible Implementation in EventStreamWriterImpl**
- Compression is done in `writeEventInternal` on the user's thread, immediately after serialization.
- Similar heuristics as above can be applied that will decide whether the data buffer is compressible or not (i.e., min size, min entropy, etc).

### Next steps
- Consider steps to streamline them and fully integrate them into the stack (handling reads, recovery, etc.).
- Consider using a mix of Client and BK compression for best results (see benchmark results in comment below)

**Comment by andreipaduroiu** [comment]
This illustrates the effect that compression has on both random (non-compressible) and pattern (compressible) data. This was generated using the SelfTester on my computer, so all numbers are for comparison only (so that we can see the effect of compression; they should not be used to measure general Pravega performance).

Sample SelfTester args (replace `Dp`, `Dpp`, `Dws` and `Do` with other values).
```
-Dtarget=SegmentStore -Dbkc=1 -Dc=1 -Ds=1 -Dsc=1 -Dp=100 -Dpp=100 -Dws=1000 -Do=1000000 -Dreads=false
```
<table cellspacing="0" border="0">
	<colgroup></colgroup>
	<colgroup></colgroup>
	<colgroup span="2" ></colgroup>
	<colgroup></colgroup>
	<colgroup></colgroup>
	<colgroup span="3"></colgroup>
	<colgroup></colgroup>
	<tr>
		<td align="center" valign=bottom><b>Type</b></td>
		<td align="center" valign=bottom><b>Events</b></td>
		<td align="center" valign=bottom><b>Producers</b></td>
		<td align="center" valign=bottom><b>Tput</b></td>
		<td align="center" valign=bottom><b>L_avg</b></td>
		<td align="center" valign=bottom><b>L50</b></td>
		<td align="center" valign=bottom><b>L75</b></td>
		<td align="center" valign=bottom><b>L90</b></td>
		<td align="center" valign=bottom><b>L99</b></td>
		<td align="center" valign=bottom><b>L999</b></td>
	</tr>
	<tr>
		<td  align="left" valign=bottom>Base</td>
		<td align="right" valign=bottom>100B</td>
		<td align="right" valign=bottom>100x100</td>
		<td align="right" valign=bottom >27.6</td>
		<td align="right" valign=bottom>31</td>
		<td align="right" valign=bottom>28</td>
		<td align="right" valign=bottom>31</td>
		<td align="right" valign=bottom >38</td>
		<td align="right" valign=bottom >157</td>
		<td align="right" valign=bottom >259</td>
	</tr>
	<tr>
		<td align="left" valign=bottom>Compress.Random</td>
		<td align="right" valign=bottom>100B</td>
		<td align="right" valign=bottom>100x100</td>
		<td align="right" valign=bottom >27.1</td>
		<td align="right" valign=bottom >32</td>
		<td align="right" valign=bottom >29</td>
		<td align="right" valign=bottom>31</td>
		<td align="right" valign=bottom >39</td>
		<td align="right" valign=bottom >175</td>
		<td align="right" valign=bottom>290</td>
	</tr>
	<tr>
		<td align="left" valign=bottom>Compress.Pattern</td>
		<td align="right" valign=bottom>100B</td>
		<td align="right" valign=bottom>100x100</td>
		<td align="right" valign=bottom >27.6</td>
		<td align="right" valign=bottom>32</td>
		<td align="right" valign=bottom >29</td>
		<td align="right" valign=bottom>31</td>
		<td align="right" valign=bottom >41</td>
		<td align="right" valign=bottom >162</td>
		<td align="right" valign=bottom >272</td>
	</tr>
	<tr>
		<td align="left" valign=bottom><br></td>
		<td align="right" valign=bottom><br></td>
		<td align="right" valign=bottom><br></td>
		<td align="left" valign=bottom><br></td>
		<td align="left" valign=bottom><br></td>
		<td align="left" valign=bottom><br></td>
		<td align="left" valign=bottom><br></td>
		<td align="left" valign=bottom><br></td>
		<td align="left" valign=bottom><br></td>
		<td align="left" valign=bottom><br></td>
	</tr>
	<tr>
		<td align="left" valign=bottom>Base</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom>1KB</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom>100x100</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom >75.6</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom >150</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom >81</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom>130</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom>321</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom >1500</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom >2089</td>
	</tr>
	<tr>
		<td align="left" valign=bottom>Compress.Random</td>
		<td align="right" valign=bottom>1KB</td>
		<td align="right" valign=bottom>100x100</td>
		<td align="right" valign=bottom >75.2</td>
		<td align="right" valign=bottom >149</td>
		<td align="right" valign=bottom >79</td>
		<td align="right" valign=bottom >112</td>
		<td align="right" valign=bottom >315</td>
		<td align="right" valign=bottom >1600</td>
		<td align="right" valign=bottom >2249</td>
	</tr>
	<tr>
		<td align="left" valign=bottom>Compress.Pattern</td>
		<td align="right" valign=bottom>1KB</td>
		<td align="right" valign=bottom>100x100</td>
		<td align="right" valign=bottom >83.6</td>
		<td align="right" valign=bottom >106</td>
		<td align="right" valign=bottom >71</td>
		<td align="right" valign=bottom >105</td>
		<td align="right" valign=bottom >277</td>
		<td align="right" valign=bottom >319</td>
		<td align="right" valign=bottom >346</td>
	</tr>
	<tr>
		<td align="left" valign=bottom><br></td>
		<td align="right" valign=bottom><br></td>
		<td align="right" valign=bottom><br></td>
		<td align="left" valign=bottom><br></td>
		<td align="left" valign=bottom><br></td>
		<td align="left" valign=bottom><br></td>
		<td align="left" valign=bottom><br></td>
		<td align="left" valign=bottom><br></td>
		<td align="left" valign=bottom><br></td>
		<td align="left" valign=bottom><br></td>
	</tr>
	<tr>
		<td align="left" valign=bottom>Base</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom>10KB</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom>100x100</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom >101.2</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom >1021</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom >935</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom >1041</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom >1191</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom >4485</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom >4824</td>
	</tr>
	<tr>
		<td align="left" valign=bottom>Compress.Random</td>
		<td align="right" valign=bottom>10KB</td>
		<td align="right" valign=bottom>100x100</td>
		<td align="right" valign=bottom >101.5</td>
		<td align="right" valign=bottom >814</td>
		<td align="right" valign=bottom >836</td>
		<td align="right" valign=bottom >987</td>
		<td align="right" valign=bottom >1053</td>
		<td align="right" valign=bottom >1486</td>
		<td align="right" valign=bottom >1953</td>
	</tr>
	<tr>
		<td align="left" valign=bottom>Compress.Pattern</td>
		<td align="right" valign=bottom>10KB</td>
		<td align="right" valign=bottom>100x100</td>
		<td align="right" valign=bottom >172.8</td>
		<td align="right" valign=bottom >506</td>
		<td align="right" valign=bottom >479</td>
		<td align="right" valign=bottom >639</td>
		<td align="right" valign=bottom >692</td>
		<td align="right" valign=bottom >818</td>
		<td align="right" valign=bottom >918</td>
	</tr>
	<tr>
		<td align="left" valign=bottom><br></td>
		<td align="right" valign=bottom><br></td>
		<td align="right" valign=bottom><br></td>
		<td align="left" valign=bottom><br></td>
		<td align="left" valign=bottom><br></td>
		<td align="left" valign=bottom><br></td>
		<td align="left" valign=bottom><br></td>
		<td align="left" valign=bottom><br></td>
		<td align="left" valign=bottom><br></td>
		<td align="left" valign=bottom><br></td>
	</tr>
	<tr>
		<td align="left" valign=bottom>Base</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom>100KB</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom>100x10</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom >101.3</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom >803</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom >839</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom >974</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom >1031</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom >1332</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom >1769</td>
	</tr>
	<tr>
		<td align="left" valign=bottom>Compress.Random</td>
		<td align="right" valign=bottom>100KB</td>
		<td align="right" valign=bottom>100x10</td>
		<td align="right" valign=bottom >99.7</td>
		<td align="right" valign=bottom >837</td>
		<td align="right" valign=bottom >888</td>
		<td align="right" valign=bottom >1017</td>
		<td align="right" valign=bottom >1070</td>
		<td align="right" valign=bottom >1369</td>
		<td align="right" valign=bottom >1794</td>
	</tr>
	<tr>
		<td align="left" valign=bottom>Compress.Pattern</td>
		<td align="right" valign=bottom>100KB</td>
		<td align="right" valign=bottom>100x10</td>
		<td align="right" valign=bottom >199.1</td>
		<td align="right" valign=bottom >429</td>
		<td align="right" valign=bottom >366</td>
		<td align="right" valign=bottom >567</td>
		<td align="right" valign=bottom >628</td>
		<td align="right" valign=bottom >805</td>
		<td align="right" valign=bottom >823</td>
	</tr>
	<tr>
		<td align="left" valign=bottom><br></td>
		<td align="right" valign=bottom><br></td>
		<td align="right" valign=bottom><br></td>
		<td align="left" valign=bottom><br></td>
		<td align="left" valign=bottom><br></td>
		<td align="left" valign=bottom><br></td>
		<td align="left" valign=bottom><br></td>
		<td align="left" valign=bottom><br></td>
		<td align="left" valign=bottom><br></td>
		<td align="left" valign=bottom><br></td>
	</tr>
	<tr>
		<td align="left" valign=bottom>Base</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom>1MB</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom>10x10</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom >104.5</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom >896</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom >726</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom >858</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom >957</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom >4659</td>
		<td style="border-top: 1px solid #000000" align="right" valign=bottom >4904</td>
	</tr>
	<tr>
		<td align="left" valign=bottom>Compress.Random</td>
		<td align="right" valign=bottom>1MB</td>
		<td align="right" valign=bottom>10x10</td>
		<td align="right" valign=bottom >106.1</td>
		<td align="right" valign=bottom >866</td>
		<td align="right" valign=bottom >708</td>
		<td align="right" valign=bottom >863</td>
		<td align="right" valign=bottom >954</td>
		<td align="right" valign=bottom >4587</td>
		<td align="right" valign=bottom >4765</td>
	</tr>
	<tr>
		<td align="left" valign=bottom>Compress.Pattern</td>
		<td align="right" valign=bottom>1MB</td>
		<td align="right" valign=bottom>10x10</td>
		<td align="right" valign=bottom >192.7</td>
		<td align="right" valign=bottom >349</td>
		<td align="right" valign=bottom >313</td>
		<td align="right" valign=bottom >355</td>
		<td align="right" valign=bottom >560</td>
		<td align="right" valign=bottom >635</td>
		<td align="right" valign=bottom >648</td>
	</tr>
</table>

**Comment by RaulGracia** [comment]
That's very interesting @andreipaduroiu. Just one point: the type of data generated may be more important that it may seem, specially for some type of compressors. I wrote a paper on that specific issue some time back, maybe you can have a look to it: https://www.usenix.org/system/files/conference/fast15/fast15-paper-gracia-tinedo.pdf
Maybe we can put some more effort con the content generation of the SelfTest tool (and also in Pravega Benchmark) if compression is going to come into play in the future.

**Comment by andreipaduroiu** [comment]
@RaulGracia Thanks for the input. I'll take a look at your paper too (funny how I read a paper published in the same place around the same time for this experiment!). 

For my test I used the default GZip compressor in Java, since I only wanted to see what effect this might have. So far this looks very promising (IMO), so if we decide to add this capability we can put more thought into this.

I've played a bit more with this and added client-side compression too and updated the above comment with the results (I broke them down by BK only, Client only and BK & Client compression). They are even more interesting than I thought originally and I've highlighted a possible way to mix-and-match the two for best results.
